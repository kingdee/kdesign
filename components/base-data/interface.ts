import { tuple } from '../_utils/type'

export const Modes = tuple('multiple', 'single')
export type Mode = typeof Modes[number]

export type SelectValue = string | string[] | number | number[] | IOptionsProps[] | null

export interface IAdvancedSelectorProps<T = SelectValue> {
  prefixCls?: string // 样式前缀
  className?: string // 样式名
  children?: React.ReactNode
  style?: React.CSSProperties // 内联样式
  disabled?: boolean // 禁用
  dropdownStyle?: React.CSSProperties // 下拉样式
  notFoundContent?: React.ReactNode // 列表为空时文本
  placeholder?: string // 选择框默认文字
  showFrequent?: boolean // 是否显示历史记录与我的收藏面板
  showCollectIcon?: boolean // 是否显示收藏按钮
  showDetailIcon?: boolean // 是否显示详情按钮
  suffixIcon?: React.ReactNode // 选择框后缀图标
  mode?: Mode // 模式
  value?: IOptionsProps[] | IOptionsProps // 指定选中的条目
  columns?: [] // 下拉列表的展示列
  options?: [] // 下拉列表的数据
  size?: number // 下拉列表显示数量
  optionLabelProp?: string // 回填到选择框的属性
  dropdownFooterRender?: React.ReactNode // 自定义下拉框footer内容
  visible?: boolean
  collectList?: [] // 收藏列表
  historyList?: [] // 我的历史
  loading?: boolean // 加载中状态
  searchField?: string // 搜索字段 默认 all（全部字段）
  onChange?: () => void // 选中option调用此函数
  onSearch?: (val: string) => void // input的value变化时调用
  onSelect?: (value: T extends (infer I)[] ? I : T, option: React.ReactElement<any>) => void
  onShowDetail?: (value: IOptionsProps[]) => void
  onShowMore?: () => void
  onCollect?: (flag: boolean, value: IOptionsProps) => void
}

export interface IIndxPosListProps {
  end: number
  start: number
}

export interface IColumnsProps {
  title?: string
  key: string
}

export interface IOptionsProps {
  value?: string | number
  label?: string
}

export interface ISearchInfoProps {
  searchValue?: string
  previousInputValue: string
  previousEditValue?: string
  searchIndex?: number
  selectionStart?: number
  deleteEndIndx: number
  editOptions?: any[]
}
