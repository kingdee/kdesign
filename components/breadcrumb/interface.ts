import { ReactNode } from 'react'
import { DropDownProps } from '../dropdown'
import { tuple } from '../_utils/type'

export const ColorModelTypes = tuple('emphasize', 'weaken')
export type ColorModelType = typeof ColorModelTypes[number]

export interface IBreadcrumbProps {
  style?: Record<string, unknown> // 内联样式
  className?: string // 类名
  colorModel?: ColorModelType // 颜色模式，默认是 'emphasize'
  prefixCls?: string // 样式前缀
  items: Array<IBreadcrumbItem> // 面包屑导航的各项，支持传入 ReactNode 类型
  separator?: ReactNode // 分隔符，默认是 '/'
  children?: ReactNode // 子元素
  onItemClick?: (item: IBreadcrumbItem, index: number) => void // 点击面包屑导航的回调函数
}

export interface IBreadcrumbItems {
  item: IBreadcrumbItem
  index: number
  separator?: ReactNode
  openEllipsis?: boolean
  onItemClick?: (item: IBreadcrumbItem, index: number) => void // 点击面包屑导航的回调函数
}

export interface IBreadcrumbItem {
  className?: string // 类名
  title?: ReactNode
  dropdownProps?: DropDownProps
  href?: string // 链接地址，不能和path共用
  path?: string // 拼接地址，不能和href共用
  icon?: ReactNode
}

export interface IItemsWidth {
  width: number
  index: number
}
