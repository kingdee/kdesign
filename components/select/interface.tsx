import * as React from 'react'
import { tuple } from '../_utils/type'
import { PopperProps } from '../_utils/usePopper'
export const SelectSizes = tuple('large', 'middle', 'small') // 选择框大小
export type SelectSize = typeof SelectSizes[number]

export const BorderTypes = tuple('none', 'underline', 'bordered')
export type BorderType = typeof BorderTypes[number]

export const Modes = tuple('multiple', 'single')
export type Mode = typeof Modes[number]
export interface AbstractSelectProps extends PopperProps {
  prefixCls?: string // 样式前缀
  className?: string // 样式名
  size?: SelectSize // 尺寸
  borderType?: BorderType // 边框类型
  notFoundContent?: React.ReactNode | null // 列表为空时文本
  showSearch?: boolean // 可搜索
  showArrow?: boolean // 显示下拉小箭头
  defaultOpen?: boolean // 默认是否展开下拉列表
  allowClear?: boolean // 允许清除内容
  disabled?: boolean // 禁用
  style?: React.CSSProperties // 内联样式
  placeholder?: string | React.ReactNode // 选择框默认文字
  dropdownClassName?: string // 下拉样式名
  dropdownStyle?: React.CSSProperties // 下拉样式
  getPopupContainer?: (triggerNode: HTMLElement) => HTMLElement // 菜单渲染父节点
  dropdownRender?: (menu: React.ReactElement) => React.ReactElement // 自定义下拉框渲染内容
  id?: string
  children?: React.ReactNode
  options?: OptionsType[]
  loading?: boolean // 加载中状态
  clearIcon?: React.ReactNode // 清除图标
  searchIcon?: React.ReactNode // 搜索图标
  suffixIcon?: React.ReactNode // 选择框后缀图标
  maxTagCount?: number // 最多显示tag数
  listHeight?: number // 下拉列表滚动高度
  labelInValue?: boolean // 每个选项的 label 包装到 value 中
  maxTagPlaceholder?: React.ReactNode | ((omittedValues: LabeledValue[]) => React.ReactNode) // 隐藏 tag 时显示的内容
  filterOption?: boolean | ((inputValue: string, option?: OptionsType) => boolean)
  optionFilterProp?: string
}

export interface LabeledValue {
  key?: string
  label?: React.ReactNode
}

export interface OptionsType {
  value?: string
  label?: React.ReactNode
}

export type SelectValue = string | string[] | number | number[] | LabeledValue | LabeledValue[] | null | undefined
export interface ISelectProps<T extends SelectValue> extends AbstractSelectProps {
  value?: T
  defaultValue?: T
  mode?: Mode
  autoFocus?: boolean
  onChange?: (value?: T, option?: React.ReactElement<any> | React.ReactElement<any>[]) => void
  onSelect?: (value?: T extends (infer I)[] ? I : T, option?: React.ReactElement<any>) => void
  onBlur?: (value?: T) => void
  onFocus?: () => void
  onSearch?: (value?: string) => void
  onClear?: () => void
  onDropdownVisibleChange?: (value?: T) => void
}
