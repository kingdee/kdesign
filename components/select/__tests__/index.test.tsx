import React from 'react'
import { render, mount } from 'enzyme'
import Select from '../index'
import { SelectSizes, BorderTypes, Modes, ISelectProps, SelectValue } from '../interface'
import mountTest from '../../../tests/shared/mountTest'
import ConfigProvider from '../../config-provider/index'

interface compProps {
  value: any
  defaultValue: any
  [key: string]: any
}

const testValueAndDefaultValue = (Component: any, props: compProps, extra: any) => {
  const { result, className } = extra?.value || {}
  const { result: defResult, className: defClassName } = extra?.defaultValue || {}
  const { args = [], result: newValue = [], callback } = extra?.onChange || {}
  it('should display value when both value and defaultValue exist', () => {
    const wrapper = mount(<Component {...props} />)
    expect(wrapper.find(className).text()).toBe(result)
  })
  it('should display defaultValue when only defaultValue exists', () => {
    const wrapper = mount(<Component {...props} value={undefined} />)
    expect(wrapper.find(defClassName).text()).toBe(defResult)
  })
  it('should not change value when selected in the component', () => {
    const wrapper = mount(<Component {...props} />)
    callback(wrapper)
    expect(wrapper.find(className).text()).toBe(result)
  })

  it('should change value when selected outside the component', () => {
    let changeValue = props.value
    const handleChange = jest.fn((...rest) => {
      rest.map((item, index) => {
        switch (args[index]) {
          case 'value':
            expect(item).toEqual(newValue[index])
            changeValue = item
            break
          case 'event':
            expect(item?.target?.value).toEqual(newValue[index])
            break
          default:
            break
        }
      })
    })
    const wrapper = mount(<Component {...props} value={changeValue} onChange={handleChange} />)

    callback(wrapper)
  })
}

const optionsData = [
  {
    label: '苹果',
    value: 'apple',
  },
  {
    label: '橘子',
    value: 'orange',
  },
  {
    label: '葡萄',
    value: 'grape',
  },
]

const defaultselectProps: ISelectProps<SelectValue> = {
  options: optionsData,
  onChange: jest.fn(),
  onSearch: jest.fn(),
  onSelect: jest.fn(),
  onClear: jest.fn(),
  onDeselect: jest.fn(),
  onBlur: jest.fn(),
  onFocus: jest.fn(),
}

describe('Select', () => {
  // 1.mount test
  mountTest(Select)

  // 2.render test
  it('renders correctly', () => {
    Modes.forEach((type) => {
      const wrapper = render(<Select mode={type} options={optionsData} />)
      expect(wrapper).toMatchSnapshot()
    })

    BorderTypes.forEach((type) => {
      const wrapper = render(<Select borderType={type} options={optionsData} />)
      expect(wrapper).toMatchSnapshot()
    })

    SelectSizes.forEach((type) => {
      const wrapper = render(<Select size={type} options={optionsData} />)
      expect(wrapper).toMatchSnapshot()
    })
  })

  // 3. render no child without errors
  it('render no child without errors', () => {
    expect(mount(<Select></Select>)).toMatchSnapshot()
  })

  // 4. render null or undefined without errors
  // it('render null or undefined without errors', () => {
  //   expect(
  //     mount(
  //       <Select>
  //         {null}
  //         {undefined}
  //       </Select>,
  //     ),
  //   ).toMatchSnapshot()
  // })

  // 5. displayName
  it('should have displayName static property', () => {
    const wrapper = mount(<Select></Select>)
    expect((wrapper.type() as any).displayName).toBe('Select')
  })

  // 6. class state
  it('should className or style use right', () => {
    const wrapper = mount(<Select className="my-test" style={{ color: 'red' }} data-test="test"></Select>)
    expect(wrapper.find('.kd-select')).toHaveClassName('.my-test')
    expect(wrapper.find('.kd-select')).toHaveStyle('color', 'red')
    expect(wrapper.prop('data-test')).toEqual('test')
  })

  // 7.component interaction(event)
  it('select all or unselect all when multiple Select', () => {
    let value = []
    const onChange = jest.fn((val) => {
      value = val
    })
    const { Option } = Select
    const wrapper = mount(
      <Select placeholder="请输入名称" onChange={onChange} defaultOpen={true} mode="multiple">
        <Option value="apple">苹果</Option>
        <Option value="lemon">柠檬</Option>
        <Option value="watermelon">西瓜</Option>
      </Select>,
    )
    wrapper.find('.kd-select-multiple-footer').first().find('.kd-checkbox-input').simulate('change')
    expect(onChange).toHaveBeenCalled()
    expect(value.length).toBe(3)

    const content = mount(
      <Select
        placeholder="请输入名称"
        onChange={onChange}
        defaultOpen={true}
        mode="multiple"
        defaultValue={['apple', 'lemon']}
      >
        <Option value="apple">苹果</Option>
        <Option value="lemon">柠檬</Option>
        <Option value="watermelon">西瓜</Option>
      </Select>,
    )
    content.find('.kd-select-multiple-footer').first().find('.kd-checkbox-input').simulate('change')
    expect(onChange).toHaveBeenCalled()
    expect(value.length).toBe(3)
  })

  it('onChange is not triggered when the singlle selector selects the same value', () => {
    const onChange = jest.fn()
    const { Option } = Select
    const wrapper = mount(
      <Select placeholder="请输入名称" onChange={onChange} value="apple" defaultOpen={true}>
        <Option value="apple">苹果</Option>
        <Option value="lemon">柠檬</Option>
        <Option value="watermelon">西瓜</Option>
      </Select>,
    )
    wrapper.find('.kd-select-item').at(0).simulate('click')
    expect(onChange).not.toHaveBeenCalled()
  })

  it('When the drop-down item is 0, the Select All button is not displayed', () => {
    const wrapper = mount(<Select placeholder="请输入名称" mode="multiple" defaultOpen={true}></Select>)
    expect(wrapper.find('.kd-select-multiple-footer')).not.toHaveClassName('.kd-checkbox-input')
  })

  // 8.config provider
  describe('8.config provider', () => {
    it('should config use config provider', () => {
      const localeData = {
        'Select.selectAll': '全部选中',
        'Select.seleted': '{size}项',
      }
      const selectConfig = {
        compDefaultProps: {
          Select: {
            size: 'small',
          },
        },
        localeConfig: { localeData, locale: 'zh-EN' },
      }
      const wrapper = mount(
        <ConfigProvider value={selectConfig}>
          <Select {...defaultselectProps} mode="single" />
        </ConfigProvider>,
      )
      expect(wrapper.find('.kd-select-wrapper')).toHaveClassName('.kd-select-size-small')

      const mulWrapper = mount(
        <ConfigProvider value={selectConfig}>
          <Select {...defaultselectProps} defaultOpen={true} mode="multiple" />
        </ConfigProvider>,
      )
      expect(mulWrapper.find('.kd-select-wrapper')).toHaveClassName('.kd-select-size-small')
      expect(mulWrapper.find('.kd-select-multiple-footer').first().find('.kd-checkbox-children')).toHaveText('全部选中')
      expect(mulWrapper.find('.kd-select-multiple-footer').find('.kd-select-multiple-footer-hadSelected')).toHaveText(
        '0项',
      )
    })
  })

  // 9. ref test
  describe('9. ref test', () => {
    it('should get Demo element from ref', () => {
      const ref = React.createRef() as any
      const onFocus = jest.fn()
      const onBlur = jest.fn()
      mount(<Select ref={ref} onBlur={onBlur} onFocus={onFocus}></Select>)
      expect((ref.current?.select as HTMLElement).classList.contains('kd-select')).toBe(true)
      expect(ref.current?.select?.focus).toBeTruthy()
      expect(ref.current?.select?.blur).toBeTruthy()
    })
  })

  // 10. api test
  // value & defaultValue(single&multiple)
  testValueAndDefaultValue(
    Select,
    { options: optionsData, value: 'apple', defaultValue: 'orange', defaultOpen: true },
    {
      value: { result: '苹果', className: '.kd-select-selection-item' },
      defaultValue: { result: '橘子', className: '.kd-select-selection-item' },
      onChange: {
        result: [
          'orange',
          {
            label: '橘子',
            value: 'orange',
          },
        ],
        className: '.kd-select-item',
        args: ['value', 'value'],
        callback: (wrapper: any) => wrapper.find('.kd-select-item').at(1).simulate('click'),
      },
    },
  )

  testValueAndDefaultValue(
    Select,
    { options: optionsData, value: ['apple'], defaultValue: ['orange'], defaultOpen: true, mode: 'multiple' },
    {
      value: { result: '苹果', className: '.kd-tag-ellipsis' },
      defaultValue: { result: '橘子', className: '.kd-tag-ellipsis' },
      onChange: {
        result: [
          ['apple', 'orange'],
          [
            {
              label: '苹果',
              value: 'apple',
            },
            {
              label: '橘子',
              value: 'orange',
            },
          ],
        ],
        className: '.kd-select-item',
        args: ['value', 'value'],
        callback: (wrapper: any) => wrapper.find('.kd-select-item').at(1).simulate('click'),
      },
    },
  )
  it('api test', () => {
    const singlewrapper = mount(<Select {...defaultselectProps} mode="single"></Select>)
    const multiplewrapper = mount(<Select {...defaultselectProps} mode="multiple"></Select>)

    // placeholder
    singlewrapper.setProps({ placeholder: 'kd' })
    expect(singlewrapper.find('.kd-select-wrapper .kd-select-placeholder').text()).toBe('kd')
    multiplewrapper.setProps({ placeholder: 'kd' })
    expect(multiplewrapper.find('.kd-select-wrapper .kd-select-placeholder').text()).toBe('kd')

    // allowClear clearIcon onClear
    expect(singlewrapper.find('.kd-select-icon-clear').length).toBe(0)
    expect(multiplewrapper.find('.kd-select-icon-clear').length).toBe(0)

    singlewrapper.setProps({ allowClear: true, value: 'apple', clearIcon: 'kd' })
    singlewrapper.update()
    expect(singlewrapper.find('.kd-select-icon-clear')).toHaveText('kd')
    singlewrapper.find('.kd-select-icon-clear').simulate('click')
    expect(defaultselectProps.onChange).toHaveBeenCalled()
    expect(defaultselectProps.onClear).toHaveBeenCalled()

    multiplewrapper.setProps({ allowClear: true, value: ['apple'], clearIcon: 'kd' })
    multiplewrapper.update()
    expect(multiplewrapper.find('.kd-select-icon-clear')).toHaveText('kd')
    multiplewrapper.find('.kd-select-icon-clear').simulate('click')
    expect(defaultselectProps.onChange).toHaveBeenCalled()
    expect(defaultselectProps.onClear).toHaveBeenCalled()

    // borderType
    BorderTypes.forEach((type) => {
      const singlewrapper = mount(<Select {...defaultselectProps} borderType={type} mode="single"></Select>)
      const multiplewrapper = mount(<Select {...defaultselectProps} borderType={type} mode="multiple"></Select>)
      expect(singlewrapper.find(`.kd-select-${type === 'none' ? 'borderless' : type}`).length).toBe(1)
      expect(multiplewrapper.find(`.kd-select-${type === 'none' ? 'borderless' : type}`).length).toBe(1)
    })

    // tree disabled (TODO click)
    singlewrapper.setProps({ disabled: true })
    expect(singlewrapper.find('.kd-select-single-disabled').length).toBe(1)
    multiplewrapper.setProps({ disabled: true })
    expect(multiplewrapper.find('.kd-select-multiple-disabled').length).toBe(1)

    // defaultOpen
    const defaultOpenSingle = mount(
      <Select
        {...defaultselectProps}
        defaultOpen={true}
        dropdownClassName="kd-select-class"
        dropdownStyle={{ background: 'red' }}
        mode="single"
      ></Select>,
    )
    const defaultOpenMultiple = mount(
      <Select
        {...defaultselectProps}
        defaultOpen={true}
        dropdownStyle={{ background: 'red' }}
        dropdownClassName="kd-select-class"
        mode="multiple"
      ></Select>,
    )
    expect(defaultOpenSingle.find('.kd-select-dropdown').length).toBe(1)
    expect(defaultOpenMultiple.find('.kd-select-dropdown').length).toBe(1)

    // dropdownClassName
    expect(defaultOpenSingle.find('.kd-select-class').length).toBe(1)
    expect(defaultOpenMultiple.find('.kd-select-class').length).toBe(1)

    // dropdownStyle
    expect(defaultOpenSingle.find('.kd-select-dropdown')).toHaveStyle('background', 'red')
    expect(defaultOpenMultiple.find('.kd-select-dropdown')).toHaveStyle('background', 'red')

    // showSearch onSearch optionFilterProp（TODO）
    defaultOpenSingle.setProps({ showSearch: true })
    defaultOpenSingle.update()
    expect(defaultOpenSingle.find('.kd-select-selection-search-input').length).toBe(1)
    defaultOpenSingle.find('.kd-select-selection-search-input').simulate('change', { target: { value: '苹果' } })
    expect(defaultselectProps.onSearch).toHaveBeenCalled()
    // expect(defaultOpenSingle.find('.kd-select-item-option').length).toBe(1)

    expect(defaultOpenMultiple.find('.kd-select-selection-search-input').length).toBe(1)
    defaultOpenMultiple.find('.kd-select-selection-search-input').simulate('change', { target: { value: '苹果' } })
    expect(defaultselectProps.onSearch).toHaveBeenCalled()
    // expect(defaultOpenMultiple.find('.kd-select-item-option').length).toBe(1)

    // suffixIcon
    defaultOpenSingle.setProps({ suffixIcon: 'kd' })
    expect(defaultOpenSingle.find('.kd-select-icon-arrow')).toHaveText('kd')
    defaultOpenMultiple.setProps({ suffixIcon: 'kd' })
    expect(defaultOpenMultiple.find('.kd-select-icon-arrow')).toHaveText('kd')

    // showArrow
    expect(defaultOpenSingle.find('.kd-select-icon-arrow').length).toBe(1)
    defaultOpenSingle.setProps({ showArrow: false })
    expect(defaultOpenSingle.find('.kd-select-icon-arrow').length).toBe(0)

    expect(defaultOpenMultiple.find('.kd-select-icon-arrow').length).toBe(1)
    defaultOpenMultiple.setProps({ showArrow: false })
    expect(defaultOpenMultiple.find('.kd-select-icon-arrow').length).toBe(0)

    // onChange onSelect onDeselect
    let selectValue: any = ''
    const onChange = jest.fn((v) => {
      selectValue = v
    })

    defaultOpenSingle.setProps({ onChange, optionLabelProp: 'value' })
    defaultOpenSingle.find('.kd-select-item').at(0).simulate('click')
    expect(onChange).toHaveBeenCalled()
    expect(selectValue).toBe('apple')
    expect(defaultOpenSingle.find('.kd-select-selection-item')).toHaveText('apple')
    expect(defaultselectProps.onSelect).toHaveBeenCalled()

    defaultOpenMultiple.setProps({ onChange, optionLabelProp: 'value' })
    defaultOpenMultiple.find('.kd-select-item').at(0).simulate('click')
    expect(onChange).toHaveBeenCalled()
    expect(selectValue.toString()).toBe('apple')
    expect(defaultOpenMultiple.find('.kd-tag-ellipsis')).toHaveText('apple')
    expect(defaultselectProps.onSelect).toHaveBeenCalled()
    defaultOpenMultiple.find('.kd-select-item').at(2).simulate('click')
    expect(defaultselectProps.onDeselect).not.toHaveBeenCalled()
    defaultOpenMultiple.find('.kd-select-item').at(2).simulate('click')
    expect(defaultselectProps.onDeselect).toHaveBeenCalled()
    defaultOpenMultiple.find('.kd-select-item').at(1).simulate('click')

    // labelInValue
    defaultOpenSingle.setProps({ labelInValue: true })
    defaultOpenSingle.find('.kd-select-item').at(1).simulate('click')
    expect(selectValue.value).toBe('orange')

    defaultOpenMultiple.setProps({ labelInValue: true })
    defaultOpenMultiple.find('.kd-select-item').at(1).simulate('click')
    defaultOpenMultiple.update()
    expect(selectValue[0].value).toBe('apple')

    // maxTagCount maxTagPlaceholder
    expect(defaultOpenMultiple.find('.kd-select-selection-tag').length).toBe(1)
    defaultOpenMultiple.setProps({
      value: ['apple', 'orange'],
      maxTagCount: 1,
      maxTagPlaceholder: <span className="kd-maxTagPlaceholder">kd</span>,
    })
    defaultOpenMultiple.update()
    expect(defaultOpenMultiple.find('.kd-select-selection-tag').length).toBe(1)
    expect(defaultOpenMultiple.find('.kd-maxTagPlaceholder')).toHaveText('kd')

    // listHeight
    defaultOpenSingle.setProps({ listHeight: 50 })
    expect(defaultOpenSingle.find('.kd-select-dropdown>div').at(0)).toHaveStyle({ maxHeight: 50 })

    defaultOpenMultiple.setProps({ listHeight: 50 })
    expect(defaultOpenMultiple.find('.kd-select-dropdown>div').at(0)).toHaveStyle({ maxHeight: 50 })

    // treeNode disabled
    defaultOpenSingle.setProps({
      options: [
        ...optionsData,
        {
          children: '柠檬',
          value: 'lemon',
          disabled: true,
        },
      ],
    })
    defaultOpenSingle.find('.kd-select-item').at(3).simulate('click')
    expect(defaultOpenSingle.find('.kd-select-item').at(3)).toHaveClassName('kd-select-item-option-disabled')

    defaultOpenMultiple.setProps({
      options: [
        ...optionsData,
        {
          children: '柠檬',
          value: 'lemon',
          disabled: true,
        },
      ],
    })
    defaultOpenMultiple.find('.kd-select-item').at(3).simulate('click')
    expect(defaultOpenMultiple.find('.kd-select-item').at(3)).toHaveClassName('kd-select-item-option-disabled')

    // dropdownRender
    defaultOpenSingle.setProps({ dropdownRender: () => <span>kd</span> })
    expect(defaultOpenSingle.find('.kd-select-dropdown div span').text()).toBe('kd')

    defaultOpenMultiple.setProps({ dropdownRender: () => <span>kd</span> })
    expect(defaultOpenMultiple.find('.kd-select-dropdown div').first().find('span').text()).toBe('kd')

    // notFoundContent
    defaultOpenSingle.setProps({ options: [], notFoundContent: 'kd' })
    expect(defaultOpenSingle.find('.kd-select-dropdown-empty')).toHaveText('kd')
    defaultOpenMultiple.setProps({ options: [], notFoundContent: 'kd' })
    expect(defaultOpenMultiple.find('.kd-select-dropdown-empty')).toHaveText('kd')

    // defaultValue
    const singleDeafWrapper = mount(<Select {...defaultselectProps} defaultValue="apple" mode="single"></Select>)
    const multipleDeafWrapper = mount(
      <Select {...defaultselectProps} defaultValue={['apple', 'orange']} mode="multiple"></Select>,
    )
    expect(singleDeafWrapper.find('.kd-select-selection-item')).toHaveText('苹果')
    expect(multipleDeafWrapper.find('.kd-tag-ellipsis').length).toBe(2)
    expect(multipleDeafWrapper.find('.kd-tag-ellipsis').at(0)).toHaveText('苹果')
    expect(multipleDeafWrapper.find('.kd-tag-ellipsis').at(1)).toHaveText('橘子')

    // filterOption（TODO）
    // const filterOption = jest.fn((searchValue, option) => {
    //   if (option.title?.indexOf(searchValue)) {
    //     return true
    //   }
    //   return false
    // })
    // singlewrapper.setProps({ disabled: false, filterOption })
    // singlewrapper.find('.kd-select-selection-search-input').simulate('change', { target: { value: '苹果' } })

    // getPopupContainer
    const wrapperRef = React.createRef() as any
    const popupContainer = mount(
      <div ref={wrapperRef}>
        <Select defaultOpen {...defaultselectProps} showSearch getPopupContainer={() => wrapperRef.current} />
      </div>,
    )

    expect(popupContainer.childAt(0).children().at(1).find('.kd-select-dropdown').length).toBe(1)

    // onFocus onBlur
    popupContainer.find('.kd-select-selection-search-input').simulate('focus')
    expect(defaultselectProps.onFocus).toBeCalled()
    popupContainer.find('.kd-select-selection-search-input').simulate('blur')
    expect(defaultselectProps.onBlur).toBeCalled()
  })
})
