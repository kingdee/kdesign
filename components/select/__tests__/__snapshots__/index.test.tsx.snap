// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select render no child without errors 1`] = `
<Select>
  <div
    className="kd-select"
  >
    <span
      className="kd-select-selector"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
    >
      <div
        className="kd-select-underline kd-select-size-middle kd-select-wrapper kd-select-single"
      >
        <span
          className="kd-select-selection-search"
        >
          <input
            className="kd-select-selection-search-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            readOnly={true}
            value=""
          />
        </span>
        <span
          className="kd-select-placeholder"
        >
          请输入需要搜索的内容
        </span>
        <span
          className="kd-select-suffix"
        >
          <span
            className="kd-select-icon-arrow kd-select-icon-arrow-down"
          >
            <Icon
              type="arrow-down"
            >
              <i
                className="kdicon kdicon-arrow-down"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </span>
        </span>
      </div>
    </span>
  </div>
</Select>
`;

exports[`Select renders correctly 1`] = `
<div
  class="kd-select"
>
  <span
    class="kd-select-selector"
    tabindex="0"
  >
    <div
      class="kd-select-underline kd-select-size-middle kd-select-wrapper kd-select-show-search kd-select-multiple kd-select-placeholder"
    >
      <span
        class="kd-select-selection-search"
        style="width:4px"
      >
        <input
          class="kd-select-selection-search-input"
          value=""
        />
        <span
          class="kd-select-selection-search-mirror"
        >
           
        </span>
      </span>
      <span
        class="kd-select-placeholder"
      >
        请输入需要搜索的内容
      </span>
      <span
        class="kd-select-suffix"
      >
        <span
          class="kd-select-icon-arrow kd-select-icon-arrow-down"
        >
          <i
            class="kdicon kdicon-arrow-down"
          />
        </span>
      </span>
    </div>
  </span>
</div>
`;

exports[`Select renders correctly 2`] = `
<div
  class="kd-select"
>
  <span
    class="kd-select-selector"
    tabindex="0"
  >
    <div
      class="kd-select-underline kd-select-size-middle kd-select-wrapper kd-select-single"
    >
      <span
        class="kd-select-selection-search"
      >
        <input
          class="kd-select-selection-search-input"
          readonly=""
          value=""
        />
      </span>
      <span
        class="kd-select-placeholder"
      >
        请输入需要搜索的内容
      </span>
      <span
        class="kd-select-suffix"
      >
        <span
          class="kd-select-icon-arrow kd-select-icon-arrow-down"
        >
          <i
            class="kdicon kdicon-arrow-down"
          />
        </span>
      </span>
    </div>
  </span>
</div>
`;

exports[`Select renders correctly 3`] = `
<div
  class="kd-select"
>
  <span
    class="kd-select-selector"
    tabindex="0"
  >
    <div
      class="kd-select-borderless kd-select-size-middle kd-select-wrapper kd-select-single"
    >
      <span
        class="kd-select-selection-search"
      >
        <input
          class="kd-select-selection-search-input"
          readonly=""
          value=""
        />
      </span>
      <span
        class="kd-select-placeholder"
      >
        请输入需要搜索的内容
      </span>
      <span
        class="kd-select-suffix"
      >
        <span
          class="kd-select-icon-arrow kd-select-icon-arrow-down"
        >
          <i
            class="kdicon kdicon-arrow-down"
          />
        </span>
      </span>
    </div>
  </span>
</div>
`;

exports[`Select renders correctly 4`] = `
<div
  class="kd-select"
>
  <span
    class="kd-select-selector"
    tabindex="0"
  >
    <div
      class="kd-select-underline kd-select-size-middle kd-select-wrapper kd-select-single"
    >
      <span
        class="kd-select-selection-search"
      >
        <input
          class="kd-select-selection-search-input"
          readonly=""
          value=""
        />
      </span>
      <span
        class="kd-select-placeholder"
      >
        请输入需要搜索的内容
      </span>
      <span
        class="kd-select-suffix"
      >
        <span
          class="kd-select-icon-arrow kd-select-icon-arrow-down"
        >
          <i
            class="kdicon kdicon-arrow-down"
          />
        </span>
      </span>
    </div>
  </span>
</div>
`;

exports[`Select renders correctly 5`] = `
<div
  class="kd-select"
>
  <span
    class="kd-select-selector"
    tabindex="0"
  >
    <div
      class="kd-select-bordered kd-select-size-middle kd-select-wrapper kd-select-single"
    >
      <span
        class="kd-select-selection-search"
      >
        <input
          class="kd-select-selection-search-input"
          readonly=""
          value=""
        />
      </span>
      <span
        class="kd-select-placeholder"
      >
        请输入需要搜索的内容
      </span>
      <span
        class="kd-select-suffix"
      >
        <span
          class="kd-select-icon-arrow kd-select-icon-arrow-down"
        >
          <i
            class="kdicon kdicon-arrow-down"
          />
        </span>
      </span>
    </div>
  </span>
</div>
`;

exports[`Select renders correctly 6`] = `
<div
  class="kd-select"
>
  <span
    class="kd-select-selector"
    tabindex="0"
  >
    <div
      class="kd-select-underline kd-select-size-large kd-select-wrapper kd-select-single"
    >
      <span
        class="kd-select-selection-search"
      >
        <input
          class="kd-select-selection-search-input"
          readonly=""
          value=""
        />
      </span>
      <span
        class="kd-select-placeholder"
      >
        请输入需要搜索的内容
      </span>
      <span
        class="kd-select-suffix"
      >
        <span
          class="kd-select-icon-arrow kd-select-icon-arrow-down"
        >
          <i
            class="kdicon kdicon-arrow-down"
          />
        </span>
      </span>
    </div>
  </span>
</div>
`;

exports[`Select renders correctly 7`] = `
<div
  class="kd-select"
>
  <span
    class="kd-select-selector"
    tabindex="0"
  >
    <div
      class="kd-select-underline kd-select-size-middle kd-select-wrapper kd-select-single"
    >
      <span
        class="kd-select-selection-search"
      >
        <input
          class="kd-select-selection-search-input"
          readonly=""
          value=""
        />
      </span>
      <span
        class="kd-select-placeholder"
      >
        请输入需要搜索的内容
      </span>
      <span
        class="kd-select-suffix"
      >
        <span
          class="kd-select-icon-arrow kd-select-icon-arrow-down"
        >
          <i
            class="kdicon kdicon-arrow-down"
          />
        </span>
      </span>
    </div>
  </span>
</div>
`;

exports[`Select renders correctly 8`] = `
<div
  class="kd-select"
>
  <span
    class="kd-select-selector"
    tabindex="0"
  >
    <div
      class="kd-select-underline kd-select-size-small kd-select-wrapper kd-select-single"
    >
      <span
        class="kd-select-selection-search"
      >
        <input
          class="kd-select-selection-search-input"
          readonly=""
          value=""
        />
      </span>
      <span
        class="kd-select-placeholder"
      >
        请输入需要搜索的内容
      </span>
      <span
        class="kd-select-suffix"
      >
        <span
          class="kd-select-icon-arrow kd-select-icon-arrow-down"
        >
          <i
            class="kdicon kdicon-arrow-down"
          />
        </span>
      </span>
    </div>
  </span>
</div>
`;
